### PR TITLE
http: http/1 codec refactor

### PR DESCRIPTION
**Pull request by mattklein123**
These changes are required for fully splitting the header map into
discrete types.
1) Prepare for concrete types
2) Fully remove client pipelining support as we don't actually use it
   anywhere and it made the refactor more complicated.
3) As a bonus, remove per-request/response heap allocations for HTTP/1.

Risk Level: Medium
Testing: Existing tests (had to modify one, see my comment in the review)
Docs Changes: N/A
Release Notes: N/A

### CODE REVIEW

File: source/common/http/http1/codec_impl.cc [hunk 1]
```diff
@@ -693,26 +683,38 @@ void ServerConnectionImpl::handlePath(HeaderMap& headers, unsigned int method) {
...
   // Handle the case where response happens prior to request complete. It's up to upper layer code
   // to disconnect the connection but we shouldn't fire any more events since it doesn't make
   // sense.
-  if (active_request_) {
+  if (active_request_.has_value()) {
     auto& headers = absl::get<RequestHeaderMapPtr>(headers_or_trailers_);
     ENVOY_CONN_LOG(trace, "Server: onHeadersComplete size={}", connection_, headers->size());
     const char* method_string = http_method_str(static_cast<http_method>(parser_.method));
 
+    if (!handling_upgrade_ && connection_header_sanitization_ && headers->Connection()) {
```
Comment (mattklein123) [inline source/common/http/http1/codec_impl.cc]: This is a slight functional change. In the old code, sanitizing never happened if there was an upgrade attempt in the request. Now it happens if the upgrade was cleaned of h2c. I did this for simplicity because I can't see any reason for the previous behavior. Please let me know if there is something I am missing.

File: test/common/http/http1/codec_impl_test.cc [hunk 2]
```diff
@@ -1155,7 +1155,7 @@ TEST_F(Http1ServerConnectionImplTest, IgnoreUpgradeH2cCloseEtc) {
   TestHeaderMapImpl expected_headers{{":authority", "www.somewhere.com"},
                                      {":path", "/"},
                                      {":method", "GET"},
-                                     {"connection", "Close,Etc"}};
```
Comment (mattklein123) [inline test/common/http/http1/codec_impl_test.cc]: This change is due to the prod code change I pointed out above. Let me know what you think.

**Comment by mattklein123** [comment]
@asraa @alyssawilk PTAL thank you!

File: source/common/http/http1/codec_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: cc @yanavlasov also who I think was heavily involved in the original review on the connection sanitizing PR.

File: source/common/http/http1/codec_impl.cc [hunk 3]
```diff
@@ -562,16 +562,6 @@ int ConnectionImpl::onHeadersCompleteBase() {
       ENVOY_CONN_LOG(trace, "codec entering upgrade mode.", connection_);
```
Comment (asraa) [inline source/common/http/http1/codec_impl.cc]: Gotcha, perhaps the PR author assumed full connection sanitization was happening in the h2c case, but it only removed upgrade/http2settings from connection header. Speaking of, can you make the quick change from
`current_headers.remove(Headers::get().Http2Settings);` 
`current_headers.removeHttp2Settings();`

File: test/common/http/http1/codec_impl_test.cc [hunk 4]
```diff
@@ -1352,24 +1352,12 @@ TEST_F(Http1ClientConnectionImplTest, FlowControlReadDisabledReenable) {
...
   ON_CALL(connection_, write(_, _)).WillByDefault(AddBufferToString(&output));
 
-  // 1st pipeline request.
+  // Request.
   TestRequestHeaderMapImpl headers{{":method", "GET"}, {":path", "/"}, {":authority", "host"}};
   request_encoder->encodeHeaders(headers, true);
   EXPECT_EQ("GET / HTTP/1.1\r\nhost: host\r\ncontent-length: 0\r\n\r\n", output);
   output.clear();
 
-  // 2nd pipeline request.
```
Comment (asraa) [inline test/common/http/http1/codec_impl_test.cc]: For my own knowledge, why was this removed in this PR?

File: source/common/http/http1/codec_impl.cc (hunk 3 above)
Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: Http2Settings is not an O(1) header currently?

File: test/common/http/http1/codec_impl_test.cc (hunk 4 above)
Comment (mattklein123) [reply test/common/http/http1/codec_impl_test.cc]: When the header maps are fully split from an interface perspective, the existing code did not work very well as we had the wrong header type at various places in the base class. This was especially problematic with HEAD requests. I could have probably fixed this a different way, but I realized that if I completely removed pipelining support, fixing became a lot easier. This also removes all heap allocations for HTTP/1 requests and responses which is a pretty nice perf benefit. Given that we don't actually use pipelining anywhere in the real code, this seemed like a good simplification to make.

BTW this is why I split this change out into its own PR because it definitely needs a good review.

File: source/common/http/http1/codec_impl.cc (hunk 3 above)
Comment (asraa) [reply source/common/http/http1/codec_impl.cc]: Oh, maybe it's not that significant, you just don't have to do the lookup at all if you use `remove##name`.

File: source/common/http/http1/codec_impl.cc [hunk 5]
```diff
@@ -387,13 +387,14 @@ void ConnectionImpl::completeLastHeader() {
   ENVOY_CONN_LOG(trace, "completed header: key={} value={}", connection_,
                  current_header_field_.getStringView(), current_header_value_.getStringView());
 
+  auto& current_headers = headersOrTrailers();
```
Comment (alyssawilk) [inline source/common/http/http1/codec_impl.cc]: this is current headers or trailers as well, no?   Can you scan through and update naming while you're in here?

File: test/common/http/http1/codec_impl_test.cc (hunk 2 above)
Comment (alyssawilk) [reply test/common/http/http1/codec_impl_test.cc]: We could summon @PiotrSikora ?

File: source/common/http/http1/codec_impl.cc (hunk 1 above)
Comment (alyssawilk) [reply source/common/http/http1/codec_impl.cc]: My guess is it was more because the upgrade case is finicky about upgrade headers, but keep-alive and upgrade are special cased so no worries there.

This is the sort of change we'd runtime guard in-house, and while I suspect it's too much of a PITA to guard the whole refactor, maybe at least call out the functional change in release notes?

File: source/common/http/http1/codec_impl.cc (hunk 5 above)
Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: Yes will do.

File: source/common/http/http1/codec_impl.cc (hunk 1 above)
Comment (mattklein123) [reply source/common/http/http1/codec_impl.cc]: Sure will add a release note. IMO this change is actually safer to some extent security wise so I would call it an improvement.

**Comment by mattklein123** [comment]
@alyssawilk updated PTAL
